trackpy/artificial.py

```diff
@@ -0,0 +1,180 @@
+from __future__ import (absolute_import, division, print_function,
+                        unicode_literals)
+import six
+import numpy as np
+from scipy.spatial import cKDTree
+from trackpy.utils import validate_tuple
+
+
+def draw_point(image, pos, value):
+    image[tuple(pos)] = value
+
+
+def feat_gauss(r, rg=0.333):
+    """ Gaussian at r = 0 with max value of 1. Its radius of gyration is
+    given by rg. """
+    return np.exp((r/rg)**2 * r.ndim/-2)
+
+
+def feat_gauss_edge(r, value_at_edge=0.1):
+    """ Gaussian at r = 0 with max value of 1. Its value at r = 1 is given by
+    value_at_edge. """
+    return np.exp(np.log(value_at_edge)*r**2)
+
+
+def feat_ring(r, r_at_max, value_at_edge=0.1):
+    """ Ring feature with a gaussian profile, centered at r_at_max. Its value
+    at r = 1 is given by value_at_edge."""
+    return np.exp(np.log(value_at_edge)*((r - r_at_max) / (1 - r_at_max))**2)
+
+
+def feat_hat(r, disc_size, value_at_edge=0.1):
+    """ Solid disc of size disc_size, with Gaussian smoothed borders. """
+    mask = r > disc_size
+    spot = (~mask).astype(r.dtype)
+    spot[mask] = feat_ring(r[mask], disc_size, value_at_edge)
+    spot[~mask] = 1
+    return spot
+
+
+def feat_step(r):
+    """ Solid disc. """
+    return r <= 1
+
+
+def draw_feature(image, position, diameter, max_value=None,
+                 feat_func=feat_gauss, ecc=None, **kwargs):
+    """ Draws a radial symmetric feature and adds it to the image at given
+    position.
+
+    Parameters
+    ----------
+    image : ndarray
+        image to draw features on
+    position : iterable
+        coordinates of feature position
+    diameter : number
+        defines the box that will be drawn on
+    max_value : number
+        maximum feature value. should be much less than the max value of the
+        image dtype, to avoid pixel wrapping at overlapping features
+    feat_func : function. Default: feat_gauss
+        function f(r) that takes an ndarray of radius values
+        and returns intensity values <= 1
+    ecc : positive number, optional
+        eccentricity of feature, defined only in 2D. Identical to setting
+        diameter to (diameter / (1 - ecc), diameter * (1 - ecc))
+    kwargs : keyword arguments are passed to feat_func
+    """
+    if len(position) != image.ndim:
+        raise ValueError("Number of position coordinates should match image"
+                         " dimensionality.")
+    diameter = validate_tuple(diameter, image.ndim)
+    if ecc is not None:
+        if len(diameter) != 2:
+            raise ValueError("Eccentricity is only defined in 2 dimensions")
+        if diameter[0] != diameter[1]:
+            raise ValueError("Diameter is already anisotropic; eccentricity is"
+                             " not defined.")
+        diameter = (diameter[0] / (1 - ecc), diameter[1] * (1 - ecc))
+    radius = tuple([(d - 1) / 2 for d in diameter])
```

Comment:
Now I'm really confused :wink:. Here's my idealized picture of how a symmetric 2D blob should work:

1. The Gaussian blob is a continuous function defined on (real numbers)^2, centered on an arbitrary real-valued location $x, y$. It has an Rg of `SIZE` units.
- That function is discretized. In a camera, each pixel is a (possibly weighted) average of the function over a certain region. For the artificial features, right now we sample the function at a single point, presumably the center of the pixel.
- The `locate` function tries to recover x, y, and Rg.

How is this different from what we are doing now? Is the problem at step 2? Are the various `+ 1` and `- 1` attempts to compensate for that?

What really troubles me is that the correction of 0.2 is approximately the same for different sizes. As the Gaussian gets larger (and therefore smoother), the details of the discretization should matter less and less. So why is it constant?